src/aleph/web/controllers/storage.py

```diff
@@ -114,59 +117,168 @@ class StorageMetadata(pydantic.BaseModel):
     sync: bool
 
 
-class UploadedFile(Protocol):
+class UploadedFile:
+    def __enter__(self):
+        raise NotImplementedError
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.cleanup()
+
+    def cleanup(self):
+        pass
+
+    async def read_and_validate(self):
+        pass
+
     @property
     def size(self) -> int:
-        ...
+        raise NotImplementedError
 
     @property
-    def content(self) -> Union[str, bytes]:
-        ...
+    def content(self) -> bytes:
+        raise NotImplementedError
 
+    @property
+    def file(self):
+        raise NotImplementedError
+
+    def get_hash(self) -> str:
+        raise NotImplementedError
 
-class MultipartUploadedFile:
-    _content: Optional[bytes]
-    size: int
 
-    def __init__(self, file_field: FileField, size: int):
+class MultipartUploadedFile(UploadedFile):
+    def __init__(self, file_field: BodyPartReader, max_size: int):
         self.file_field = file_field
-        self.size = size
-        self._content = None
+        self.max_size = max_size
+        self._hash = None
+        try:
+            self._temp_file = tempfile.NamedTemporaryFile(delete=False)
+            self._file_content = bytearray()
+        except Exception as e:
+            web.HTTPInternalServerError(reason="Cannot create tempfile")
+
+    def __enter__(self):
+        self._temp_file.seek(0)
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        try:
+            self._temp_file.close()
+            if os.path.exists(self._temp_file.name):
+                os.unlink(self._temp_file.name)
+        except Exception as e:
+            web.HTTPInternalServerError(reason="Cannot create tempfile")
+
+    async def read_and_validate(self):
+        total_read = 0
+        chunk_size = 8192
+        hash_sha256 = hashlib.sha256()
+
+        while total_read < self.max_size:
+            chunk = await self.file_field.read_chunk(chunk_size)
```

Comment:
You probably want to do an "async for" like you did here https://github.com/aleph-im/pyaleph/pull/565/files#diff-3ac5ca0d72a6758698e8f83939e455e54c6d97047c7c4310b7f9e3dbd963b2d7R222 since the same condition is being taken care of in the if line 182 https://github.com/aleph-im/pyaleph/pull/565/files#diff-3ac5ca0d72a6758698e8f83939e455e54c6d97047c7c4310b7f9e3dbd963b2d7R182

Reply:
Reading [BodyPartReader source code](https://docs.aiohttp.org/en/stable/_modules/aiohttp/multipart.html#BodyPartReader), it does support the async iterator protocol so you should be able to to rewrite this code like this:

```python
async for chunk in self.file_field.read_chunk(chunk_size):
    yield chunk
```

Edit: turns out you did already use `async for here https://github.com/aleph-im/pyaleph/pull/565/files?diff=split#diff-3ac5ca0d72a6758698e8f83939e455e54c6d97047c7c4310b7f9e3dbd963b2d7R198-R200

So maybe this method isn't really needed? You can directly use `async for` [line 154](https://github.com/aleph-im/pyaleph/pull/565/files?diff=split#diff-3ac5ca0d72a6758698e8f83939e455e54c6d97047c7c4310b7f9e3dbd963b2d7R154) and instead and remove the 3 `_read_chunks` methods to simplify code